src/verbs/queue_pair.rs

```diff
@@ -157,147 +356,202 @@ impl<'res> QueuePairBuilder<'res> {
             )
         };
 
-        Ok(QueuePair {
+        Ok(BasicQueuePair {
             qp: NonNull::new(qp).ok_or(format!("ibv_create_qp failed, {}", io::Error::last_os_error()))?,
             _phantom: PhantomData,
         })
     }
 
-    pub fn build_ex() -> () {
-        todo!();
-    }
-}
-
-impl QueuePair<'_> {
-    pub(crate) fn new<'pd>(pd: &'pd ProtectionDomain) -> Self {
-        todo!()
-    }
+    // build extended qp
+    pub fn build_ex(&mut self) -> Result<ExtendedQueuePair<'res>, String> {
+        let qp = unsafe { ibv_create_qp_ex((*self.init_attr.pd).context, &mut self.init_attr).unwrap_or(null_mut()) };
 
-    pub fn modify(&mut self, attr: &QueuePairAttribute) -> Result<(), String> {
-        // ibv_qp_attr does not impl Clone trait, so we use struct update syntax here
-        let mut qp_attr = ibv_qp_attr { ..attr.attr };
-        let ret = unsafe { ibv_modify_qp(self.qp.as_ptr(), &mut qp_attr as *mut _, attr.attr_mask.0 as _) };
-        if ret == 0 {
-            Ok(())
-        } else {
-            Err(format!("ibv_modify_qp failed, err={ret}"))
-        }
+        Ok(ExtendedQueuePair {
+            qp_ex: NonNull::new(qp).ok_or(format!("ibv_create_qp failed, {}", io::Error::last_os_error()))?,
+            _phantom: PhantomData,
+        })
     }
 }
 
 pub struct QueuePairAttribute {
     attr: ibv_qp_attr,
-    attr_mask: ibv_qp_attr_mask,
+    attr_mask: QueuePairAttributeMask,
 }
 
 impl QueuePairAttribute {
     pub fn new() -> Self {
         QueuePairAttribute {
             attr: unsafe { MaybeUninit::zeroed().assume_init() },
-            attr_mask: ibv_qp_attr_mask(0),
+            attr_mask: QueuePairAttributeMask { bits: 0 },
         }
     }
 
     // initialize attr from an existing one (this is useful when we interact with rdmacm)
-    pub fn from(attr: &ibv_qp_attr, attr_mask: u32) -> Self {
+    pub fn from(attr: &ibv_qp_attr, attr_mask: i32) -> Self {
         QueuePairAttribute {
             attr: ibv_qp_attr { ..*attr },
-            attr_mask: ibv_qp_attr_mask(attr_mask),
+            attr_mask: QueuePairAttributeMask { bits: attr_mask },
         }
     }
 
     // TODO what about the default value for qp_attr?
 
     pub fn setup_state(&mut self, state: QueuePairState) -> &mut Self {
         self.attr.qp_state = state as _;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_STATE;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::State);
```

Comment:
Using `self.attr_mask |= QueuePairAttributeMask::State` here is more concise; This also applies for the following setup methods.